akka-stream/src/main/scala/akka/stream/impl/io/InputStreamSubscriber.scala

```diff
@@ -0,0 +1,133 @@
+  }
+
+  trait InputStreamSubscriberInMessage
+
+  case class Read(length: Int) extends InputStreamSubscriberInMessage
+
+  case object Close extends InputStreamSubscriberInMessage
+
+  trait InputStreamSubscriberOutMessage
+
+  case class Bytes(bytes: ByteString) extends InputStreamSubscriberOutMessage
+
+  case class Failure(msg: String) extends InputStreamSubscriberOutMessage
+
+}
+
+/** INTERNAL API */
+private[akka] class InputStreamSubscriber(bytesWrittenPromise: Promise[Long], bufSize: Int)
+  extends akka.stream.actor.ActorSubscriber
+  with ActorLogging {
+
+  var bytesRequestedDownstream: Option[(ActorRef, Int)] = None
+
+  var bytesSentDownstream: Long = 0
+
+  var onCompleteReceived: Boolean = false
+
+  var availableChunks: Vector[ByteString] = Vector.empty[ByteString]
+
+  override protected val requestStrategy = new RequestStrategy() {
+    override def requestDemand(remainingRequested: Int): Int = bytesRequestedDownstream match {
+      case Some(_) ⇒ if (remainingRequested > 0) 0 else 1
+      case None    ⇒ bufSize - availableChunks.size - remainingRequested
+    }
+  }
+
+  def receive = {
+    case ActorSubscriberMessage.OnNext(bytes: ByteString) ⇒
+      availableChunks :+= bytes
+      tryToReply(sendRest = false)
+
+    case ActorSubscriberMessage.OnError(cause) ⇒
+      tryToReply(sendRest = true)
+      bytesWrittenPromise.failure(cause)
+      context.stop(self)
+
+    case ActorSubscriberMessage.OnComplete ⇒
+      tryToReply(sendRest = true)
+      //block context.stop until all available data will be read or stream will be closed
+      if (availableChunks.isEmpty) context.stop(self)
+      else onCompleteReceived = true
+
+    case InputStreamSubscriber.Read(length) ⇒ onRead(length)
+
+    case InputStreamSubscriber.Close        ⇒ cancel()
+  }
+
+  def availableBytes(): Int =
+    availableChunks.foldLeft(0)((acc, byteString) ⇒ acc + byteString.size)
+
+  def sendBytes(actor: ActorRef, length: Int) = {
+    bytesSentDownstream += length
+    bytesRequestedDownstream = None
+    actor ! InputStreamSubscriber.Bytes(getBytes(length))
+  }
+
+  def onRead(requestedBytes: Int) =
+    if (onCompleteReceived)
+      onReadWhenCompleteReceived(requestedBytes)
+    else if (availableBytes() >= requestedBytes)
+      sendBytes(sender, requestedBytes)
+    else if (availableChunks.size < bufSize)
+      bytesRequestedDownstream = Some((sender, requestedBytes))
+
+  def onReadWhenCompleteReceived(requestedBytes: Int) =
+    if (availableBytes() > requestedBytes)
+      sendBytes(sender, requestedBytes)
+    else {
+      sendBytes(sender, availableBytes())
+      context.stop(self)
+    }
+
+  override def postStop(): Unit = {
+    bytesWrittenPromise.trySuccess(bytesSentDownstream)
+    super.postStop()
+  }
+
+  def tryToReply(sendRest: Boolean) =
+    bytesRequestedDownstream match {
+      case Some((actor, length)) ⇒
+        if (!sendRest) {
+          if (availableBytes >= length)
+            sendBytes(actor, length)
```

Comment:
InputStream.read routinely returns less than was expected when the data source has provided some but not enough data; I think we should do the same here.